pkg/icingadb/schema.go

```diff
@@ -56,3 +58,109 @@ func CheckSchema(ctx context.Context, db *database.DB) error {
 
 	return nil
 }
+
+// ImportSchema performs an initial schema import in the db.
+//
+// This function starts by checking if at least a partial schema exists. If it is present, a nil error will be returned.
+// Otherwise, the schema file for the db's driver will be loaded from databaseSchemaDir and imported into db.
+//
+// The function returns a bool - true when an import was performed - and a potential error. If the error is not nil,
+// the bool should also be false. When (false, nil) is returned, a schema is already imported.
+func ImportSchema(
+	ctx context.Context,
+	db *database.DB,
+	databaseName string,
+	databaseSchemaDir string,
+) (bool, error) {
+	var tableSchema, schemaFileDirPart string
+	switch db.DriverName() {
+	case database.MySQL:
+		tableSchema = databaseName
+		schemaFileDirPart = "mysql"
+
+	case database.PostgreSQL:
+		tableSchema = "public"
+		schemaFileDirPart = "pgsql"
+
+	default:
+		return false, fmt.Errorf("unsupported database driver %q", db.DriverName())
+	}
+
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return false, errors.Wrap(err, "can't start database transaction")
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	var schemaTablesCount int
+	err = retry.WithBackoff(
+		ctx,
+		func(ctx context.Context) (err error) {
+			query := tx.Rebind("SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=? AND TABLE_NAME='icingadb_schema'")
+			err = tx.QueryRowxContext(ctx, query, tableSchema).Scan(&schemaTablesCount)
+			if err != nil {
+				err = database.CantPerformQuery(err, query)
+			}
+			return
+		},
+		retry.Retryable,
+		backoff.NewExponentialWithJitter(128*time.Millisecond, 1*time.Minute),
+		db.GetDefaultRetrySettings())
+	if err != nil {
+		return false, errors.Wrap(err, "can't check if database has the schema applied")
+	}
+	if schemaTablesCount > 0 {
+		// icingadb_schema table exists; schema is at least partially applied
+		return false, nil
+	}
```

Comment:
Just reading the code, `schemaTablesCount` gave me the first impression that we would try to import missing tables. I would rather use `SELECT 1` and check for `sql.ErrNoRows`, which would also not `Scan()`:

```suggestion
	err = retry.WithBackoff(
		ctx,
		func(ctx context.Context) error {
			query := tx.Rebind("SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=? AND TABLE_NAME='icingadb_schema'")
			if err := tx.QueryRowxContext(ctx, query, tableSchema).Err(); err != nil {
				// The `SELECT 1` query returns `sql.ErrNoRows` if the `icingadb_schema` table does not exist;
				// otherwise, it returns a database error.
				return database.CantPerformQuery(err, query)
			}

			return nil
		},
		retry.Retryable,
		backoff.NewExponentialWithJitter(128*time.Millisecond, 1*time.Minute),
		db.GetDefaultRetrySettings())
	if err == nil {
		// Table `icingadb_schema` exists.
		return false, nil
	}
	if !errors.Is(err, sql.ErrNoRows) {
		return false, errors.Wrap(err, "can't check if database has the schema applied")
	}
```

Reply:
> I would rather use `SELECT 1` and check for `sql.ErrNoRows`, which would also not require `Scan()`:

This does not work on PostgreSQL (haven't tested MySQL yet). While the query in question will return no rows, the `.Err()` method does return a `nil` error.

```
icingadb=# SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='public' AND TABLE_NAME='icingadb_schema';
 ?column?
----------
(0 rows)
```


```go
			query := db.Rebind("SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=? AND TABLE_NAME='icingadb_schema'")
			if err := db.QueryRowxContext(ctx, query, tableSchema).Err(); err != nil {
				// The "SELECT 1" query returns sql.ErrNoRows if the icingadb_schema table does not exist;
				// otherwise, it returns a database error.
				return database.CantPerformQuery(err, query)
			}
			return nil // <- We will end up here; verified in the debugger.
```

_(For the record, don't click "Unresolve conversation" before sending the comment, as GitHub will clean out the comment field and purge everything you have written so far. Thanks.)_